### PR TITLE
Create pcawg controls

### PR DESCRIPTION
**Pull request by GatKrieger**
Generate synthetic signatures from a signature vcf file and a db vcf file, will be incorporated to the MRD pipeline

### CODE REVIEW

File: ugvc/vcfbed/variant_annotation.py [hunk 1]
```diff
@@ -734,3 +734,29 @@ def determine_cycle_skip_status(ref: str, alt: str, flow_order: str):
...
+    motif_ref = rec.info["X_LM"][0][-1] + rec.ref + rec.info["X_RM"][0][0]
+    motif_alt = rec.info["X_LM"][0][-1] + rec.alts[0] + rec.info["X_RM"][0][0]
+    return motif_ref + ">" + motif_alt
+
+
+def get_trinuc_substitution_dist(vcf_file):
+    trinuc_sub = get_trinuc_sub_list()
+    trinuc_dict = {k: 0 for k in trinuc_sub}
+    with pysam.VariantFile(vcf_file, "rb") as infh:
+        for rec in infh.fetch():
```
Comment (kenissur) [inline ugvc/vcfbed/variant_annotation.py:759]: I suggest checking that the rec is a snp before processing it (len(rec.ref)==1 and len(rec.alts)==1 and len(rec.alts[0])==1)

File: ugvc/mrd/mrd_utils.py [hunk 2]
```diff
@@ -772,3 +777,75 @@ def concat_dataframes(dataframes: list, outfile: str, n_jobs: int = 1):
...
+        number of synthetic signatures to generate
+    output_dir: str
+        output directory
+
+    Returns
+    -------
+    synthetic_signatures: list[str]
+        list of synthetic control vcf files
+
+    """
```
Comment (kenissur) [inline ugvc/mrd/mrd_utils.py:804]: I think you should add a "random_seed:int = 0" argument and do "np.random.seed(random_seed)" in the first line, to make sure that the output is determinstic. Specifically for your test, which I suspect could fail if it's not.

File: ugvc/mrd/mrd_utils.py [hunk 3]
```diff
@@ -772,3 +777,75 @@ def concat_dataframes(dataframes: list, outfile: str, n_jobs: int = 1):
...
+    trinuc_db = get_trinuc_substitution_dist(db_vcf)
+
+    # allocate index and counter per synthetic signature
+    trinuc_dict = {}
+    for trinucsub in trinuc_signature.keys():
+        n_subs_signature = trinuc_signature[trinucsub]
+        n_subs_db = trinuc_db[trinucsub]
+        trinuc_dict[trinucsub] = {}
+        for i in range(n_synthetic_signatures):
+            f_index = f"index_to_sample_{i}"
```
Comment (kenissur) [inline ugvc/mrd/mrd_utils.py]: Why use those strings and keys and not just i?

File: ugvc/mrd/mrd_utils.py [hunk 4]
```diff
@@ -772,3 +777,75 @@ def concat_dataframes(dataframes: list, outfile: str, n_jobs: int = 1):
...
+
+    # make output directory
+    os.makedirs(output_dir, exist_ok=True)
+
+    # read db vcf file and allocate output files and file handles
+    db_fh = pysam.VariantFile(db_vcf, "rb")
+    synthetic_signatures = []
+    file_handle_list = []
+    for i in range(n_synthetic_signatures):
+        output_vcf = pjoin(output_dir, f"syn{i}.vcf.gz")
```
Comment (kenissur) [inline ugvc/mrd/mrd_utils.py]: maybe include the basename of the db and of the signature?

**Review by kenissur — APPROVED** [review]
Looks very good. I am approving but please take care of the following before merging:

1. See minor comments in line
2. How large are the test files you've added? I couldn't fetch just now due to some technicality. Make sure the files are not too large.
3. Be sure to rebase to the most recent master first, then do "Squash and merge" when closing the PR

**Review by kenissur — APPROVED** [review]
Looks very good. I am approving but please take care of the following before merging:

1. See minor comments in line
2. How large are the test files you've added? I couldn't fetch just now due to some technicality. Make sure the files are not too large.
3. Be sure to rebase to the most recent master first, then do "Squash and merge" when closing the PR

File: ugvc/vcfbed/variant_annotation.py (hunk 1 above)
Comment (GatKrieger) [reply ugvc/vcfbed/variant_annotation.py:759]: Done

File: ugvc/mrd/mrd_utils.py (hunk 2 above)
Comment (GatKrieger) [reply ugvc/mrd/mrd_utils.py:804]: Done

File: ugvc/mrd/mrd_utils.py (hunk 3 above)
Comment (GatKrieger) [reply ugvc/mrd/mrd_utils.py]: Edited

File: ugvc/mrd/mrd_utils.py (hunk 4 above)
Comment (GatKrieger) [reply ugvc/mrd/mrd_utils.py]: Done

**Pull request merged**